File: sample/src/main/java/com/afollestad/iconrequestsample/MainAdapter.java
package com.afollestad.iconrequestsample;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.afollestad.iconrequest.AppModel;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author Aidan Follestad (afollestad)
 */
class MainAdapter extends RecyclerView.Adapter<MainAdapter.MainVH> {

  interface SelectionListener {

    void onSelection(int index, AppModel app);
  }

  private List<AppModel> appsList;
  private SelectionListener listener;

  MainAdapter() {
  }

  void setAppsList(List<AppModel> appsList) {
    this.appsList = appsList;
    notifyDataSetChanged();
  }

  void update(AppModel app) {
    for (int i = 0; i < appsList.size(); i++) {
      if (app.equals(appsList.get(i))) {
        appsList.set(i, app);
        notifyItemChanged(i);
        break;
      }
    }
  }

  void setListener(SelectionListener listener) {
    this.listener = listener;
  }

  @Override
  public MainVH onCreateViewHolder(ViewGroup parent, int viewType) {
    View view = LayoutInflater.from(parent.getContext())
        .inflate(R.layout.list_item_app, parent, false);
    return new MainVH(view, this);
  }

  @Override
  public int getItemCount() {
    return appsList != null ? appsList.size() : 0;
  }

  @Override
  public void onBindViewHolder(MainVH holder, int position) {
    final AppModel app = appsList.get(position);
    app.loadIcon(holder.icon);

    holder.title.setText(app.name());
    holder.title.setAlpha(1f);
    holder.icon.setAlpha(1f);

    holder.itemView.setActivated(app.selected());
  }

  static class MainVH extends RecyclerView.ViewHolder implements View.OnClickListener {

    @BindView(R.id.icon)
    ImageView icon;
    @BindView(R.id.title)
    TextView title;
    final MainAdapter adapter;

    MainVH(View itemView, MainAdapter adapter) {
      super(itemView);
      ButterKnife.bind(this, itemView);
      this.adapter = adapter;
      itemView.setOnClickListener(this);
    }

    @Override
    public void onClick(View view) {
      AppModel appModel = adapter.appsList.get(getAdapterPosition());
      if (adapter.listener != null) {
        adapter.listener.onSelection(getAdapterPosition(), appModel);
      }
    }
  }
}

File: sample/src/main/java/com/afollestad/iconrequestsample/MainAdapter.kt
package com.afollestad.iconrequestsample

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.afollestad.iconrequest.AppModel
import com.bumptech.glide.Glide
import kotlinx.android.synthetic.main.list_item_app.view.icon
import kotlinx.android.synthetic.main.list_item_app.view.title

typealias SelectionListener = ((Int, AppModel) -> (Unit))

/** @author Aidan Follestad (afollestad) */
internal class MainAdapter : RecyclerView.Adapter<MainAdapter.MainVH>() {

  private var appsList: MutableList<AppModel> = mutableListOf()
  private var listener: SelectionListener? = null

  fun setAppsList(appsList: List<AppModel>?) {
    this.appsList = appsList?.toMutableList() ?: mutableListOf()
    notifyDataSetChanged()
  }

  fun update(app: AppModel) {
    for (i in appsList.indices) {
      if (app.code == appsList[i].code) {
        appsList[i] = app
        notifyItemChanged(i)
        break
      }
    }
  }

  fun setListener(listener: SelectionListener) {
    this.listener = listener
  }

  override fun onCreateViewHolder(
    parent: ViewGroup,
    viewType: Int
  ): MainVH {
    val view = LayoutInflater.from(parent.context)
        .inflate(R.layout.list_item_app, parent, false)
    return MainVH(view, this)
  }

  override fun getItemCount(): Int {
    return appsList.size
  }

  override fun onBindViewHolder(
    holder: MainVH,
    position: Int
  ) {
    val app = appsList[position]
    holder.bind(app)
  }

  internal class MainVH(
    itemView: View,
    private val adapter: MainAdapter
  ) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

    init {
      itemView.setOnClickListener(this)
    }

    fun bind(model: AppModel) {
      Glide.with(itemView.icon)
          .load(model)
          .into(itemView.icon)
      itemView.icon.alpha = 1f
      itemView.title.text = model.name
      itemView.title.alpha = 1f
      itemView.isActivated = model.selected
    }

    override fun onClick(view: View) {
      val appModel = adapter.appsList[adapterPosition]
      adapter.listener?.invoke(adapterPosition, appModel)
    }
  }
}
